pkg/common/flyte_url.go

```diff
@@ -0,0 +1,92 @@
+package common
+
+import (
+	"fmt"
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/admin"
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/core"
+	"regexp"
+	"strconv"
+)
+
+type IOType string
+
+const (
+	UndefinedIo IOType = ""
+	INPUT              = "i"
+	OUTPUT             = "o"
+	DECK               = "d"
+)
+
+func ParseFlyteURL(flyteURL string) (core.NodeExecutionIdentifier, *int, IOType, error) {
+	// flyteURL is of the form flyte://v1/project/domain/execution_id/node_id/attempt/[iod]
+	// where i stands for inputs.pb o for outputs.pb and d for the flyte deck
+	// If the retry attempt is missing, the io requested is assumed to be for the node instead of the task execution
+	zero := 0
+	re, err := regexp.Compile("flyte://v1/([a-zA-Z0-9_-]+)/([a-zA-Z0-9_-]+)/([a-zA-Z0-9_-]+)/([a-zA-Z0-9_-]+)(?:/([0-9]+))?/([iod])")
+	if err != nil {
+		return core.NodeExecutionIdentifier{}, &zero, UndefinedIo, err
+	}
+	re.MatchString(flyteURL)
+	matches := re.FindStringSubmatch(flyteURL)
+	if len(matches) != 7 && len(matches) != 6 {
+		return core.NodeExecutionIdentifier{}, &zero, UndefinedIo, fmt.Errorf("failed to parse flyte url, only %d matches found", len(matches))
+	}
+	proj := matches[1]
+	domain := matches[2]
+	executionID := matches[3]
+	nodeID := matches[4]
+	var attempt *int // nil means node execution, not a task execution
+	if len(matches) == 7 && matches[5] != "" {
+		a, err := strconv.Atoi(matches[5])
+		if err != nil {
+			return core.NodeExecutionIdentifier{}, &zero, UndefinedIo, fmt.Errorf("failed to parse attempt, %s", err)
+		}
+		attempt = &a
+	}
+	var ioType IOType
+	switch matches[len(matches)-1] {
```

Comment:
once you `//go:generate enumer` you wont need this switch, a simple lookup would work